### ISSUE TITLE
Suggest infix

### ISSUE THREAD

**Issue by georgefst** [ndmitchell/hlint#1119]
IME, using prefix notation for a fully-applied operator is usually a code smell. And it can often result from refactorings or using typed hole suggestions etc. (I'm kind of thinking ahead here to the day when Hlint is fully integrated with HLS)

E.g. I'd like to see:

```haskell
(+) 1 2
```

replaced with:

```haskell
1 + 2
```

I'm a little surprised that, AFAICT, this hasn't been suggested before.

PS. I wanted to check this hadn't already been implemented, which would require knowing the full list of hints which are disabled by default. It's unclear whether the list from `hlint --default` is complete. `--generate-summary` looked promising, but the UX around `hlint test` is confusing...

**Comment by ndmitchell** [comment]
My only concern is whenever I write prefix notation, its because I really mean it. I've definitely done it with `($)` but I'm not sure I've ever done it with another operator. Certainly worth trying.

There are very few disabled by default hints, and we're working on generating them all automatically and putting them on the website - patches are in flight to do that as we speak. The list in default is definitely incomplete, and isn't intended to be exhausive.

**Comment by georgefst** [comment]
> I've definitely done it with ($)

I'm curious - why?

**Comment by ndmitchell** [comment]
> I'm curious - why?

I don't remember, but I'm sure I did. I have also definitely written `(:) 1 $ rest_of_list`. But I guess that won't be flagged as infix.

**Comment by georgefst** [comment]
Unfortunately, #626 doesn't work as I expected, so `- warn: {lhs: "(?) x y", rhs: "x ? y", name: Use infix}` suggests replacing every `f x y` with ``x `f` y``, which is... less than ideal.

So, should we change the wildcards to only match actual operators (including backticked functions?)? Or is there a good reason for it working the way it currently does?

CC @yairchu

**Comment by ndmitchell** [comment]
Does indeed seem a bit odd that operator wildcards match that. I would be happy for it to change.

**Comment by georgefst** [comment]
Started looking in to this, got frustrated by the difficulty of doing `show` debugging with the GHC API, and ended up getting totally sidetracked hacking together a script with a lot of orphan (and overlapping) `Show` instances. And taking advantage of an unreleased branch of `pretty-simple`, which allows for custom post-processing, to end up with output like this (in a terminal): https://georgefst.github.io/ghc-parse-pretty-simple/

Anyway, @ndmitchell, I've just seen that you mentioned this issue in your Munihac talk, and I'm wondering if there's a better way. The slides suggest "(use pretty-print)" as a solution, but I'm not quite sure what that refers to? Do you mean GHC's `Outputtable` class for printing out the Haskell source, rather than the raw AST?

**Comment by ndmitchell** [comment]
See the function [`unsafePrettyPrint`](https://hackage.haskell.org/package/ghc-lib-parser-ex-8.10.0.16/docs/Language-Haskell-GhclibParserEx-GHC-Utils-Outputable.html#v:unsafePrettyPrint) which is what I use for most things.

**Comment by ndmitchell** [comment]
Agreed, its not great though :(

**Comment by shayne-fletcher** [comment]
For debugging, [showAstData](https://hackage.haskell.org/package/ghc/docs/GHC-Hs-Dump.html#v:showAstData) can be very useful.

**Comment by ndmitchell** [comment]
I'd been using [gshow](https://hackage.haskell.org/package/syb-0.7.1/docs/Data-Generics-Text.html#v:gshow) but that looks much better. Perhaps we need a section in contributing that lists the best of these functions.

**Comment by shayne-fletcher** [comment]
If the goal is just to dump a parse tree in the console, `-ddump-parsed-ast` will do that for you.

**Comment by georgefst** [comment]
> For debugging, [showAstData](https://hackage.haskell.org/package/ghc/docs/GHC-Hs-Dump.html#v:showAstData) can be very useful.

Aha! That's what I was looking for, thanks

**Comment by georgefst** [comment]
> Perhaps we need a section in contributing that lists the best of these functions.

Yeah, there's now a lot of very useful info in this thread that has nothing to do with the actual issue. Would be a shame for it to get lost.

FWIW, I've settled on `function pretty-parse-ghc () { ghc -ddump-parsed-ast "$1" | pretty-simple ; }` (installable with `cabal install pretty-simple -f buildexe`).

**Comment by georgefst** [comment]
> FWIW, I've settled on `function pretty-parse-ghc () { ghc -ddump-parsed-ast "$1" | pretty-simple ; }`

Even quicker with `-fno-code`. Is there an option to make GHC stop after parsing (and not even typecheck)?

**Comment by shayne-fletcher** [comment]
> I'd been using [gshow](https://hackage.haskell.org/package/syb-0.7.1/docs/Data-Generics-Text.html#v:gshow) but that looks much better. Perhaps we need a section in contributing that lists the best of these functions.

Please feel free to assign https://github.com/ndmitchell/hlint/issues/1128 to me and I'll be happy to start a section.

**Comment by ndmitchell** [comment]
I should probably figure out how to give you permissions to assign yourself... But not at a computer with my GitHub password right now.
